### PR TITLE
71 extend event

### PR DESCRIPTION
**Pull request by cubap**
addresses #74 
Still has some unhandled errors when unexpanded renders start, but it does eventually render as expected.

Draft for now, since the edges are still ragged and this will be a big pull, but feel free to comment @thehabes in case this doesn't address your cases.

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -414,7 +414,7 @@ <h4>Simple Forms</h4>
         </footer>
 
 
-        <script src="./releases/alpha-.10/deer.js" type="module"></script>
+        <script src="./js/deer.js" type="module"></script>
```
Comment (cubap) [inline index.html:417]: I'll put this back before merging. This should be our most recent release and the text version needs to change too.

File: js/deer-record.js [hunk 2]
```diff
@@ -55,6 +55,9 @@ async function renderChange(mutationsList) {
 
 export default class DeerReport {
     constructor(elem, deer = {}) {
+
+return false
+
         for (let key in DEER) {
```
Comment (cubap) [inline js/deer-record.js:61]: obviously some trimming here.

File: js/deer-render.js [hunk 3]
```diff
@@ -54,12 +44,63 @@ async function renderChange(mutationsList) {
...
+            RENDER.detectInsertions(elem)
+        }
     }
 }
 
 const RENDER = {}
+RENDER.detectInsertions = elem => {
+    let newViews = (elem.querySelectorAll(config.VIEW).length) ? elem.querySelectorAll(config.VIEW) : []
+    let newForms = (elem.querySelectorAll(config.FORM).length) ? elem.querySelectorAll(config.VIEW) : []
+    if (newForms.length) {
```
Comment (cubap) [inline js/deer-render.js:57]: length might be more than is needed here, I forget why it was done.

File: js/deer-render.js [hunk 4]
```diff
@@ -54,12 +44,63 @@ async function renderChange(mutationsList) {
...
 
+    if (typeof templateResponse.then === "function") { templateResponse.then(elem, obj, options) }
+    //Note this is deprecated for the "deer-view-rendered" event.  above.  
+    UTILS.broadcast(undefined, DEER.EVENTS.LOADED, elem, obj)
+}
 RENDER.element = function (elem, obj) {
+    UTILS.worker.addEventListener("message", event => {
+        let templ = DEER.TEMPLATES[elem.getAttribute(DEER.TEMPLATE) || (elem.getAttribute(DEER.COLLECTION) ? "list" : "json")]
+        if (event.data.action === "expanded") {
+            RENDER.applyTemplate(elem, event.data.item, templ)
```
Comment (cubap) [inline js/deer-render.js:93]: maybe pass `options` in here so we aren't checking twice

File: js/deer-utils.js [hunk 5]
```diff
@@ -105,6 +115,23 @@ export default {
             return label || noLabel
         }
     },
+    postView(entity, matchOn = ["__rerum.generatedBy", "creator"]){
```
Comment (cubap) [inline js/deer-utils.js:118]: This is the model for form handling and other messages sent into the worker as well.

File: js/deer-utils.js [hunk 6]
```diff
@@ -105,6 +115,23 @@ export default {
...
+        let UTILS = this
```
Comment (cubap) [inline js/deer-utils.js:119]: why is this necessary?

File: js/entities.js [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+     * Take a known object with an id and query for annotations targeting it.
+     * Discovered annotations are asserted on the original object and returned.
+     * @param {Object} entity Target object to search for description
+     */
+function expandEntity(entity={}, matchOn = ["__rerum.generatedBy", "creator"]) {
+    let findId = entity["@id"] || entity.id || entity
+    if (typeof findId !== "string") {
+        return Promise.resolve(entity)
+    }
+    return fetch(findId).then(response => response.json())
```
Comment (cubap) [inline js/entities.js:12]: unchanged, just lifted from `expand()`

File: js/entities.js [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+            })
+}
+
+/**
+     * Execute query for any annotations in RERUM which target the
+     * id passed in. Promise resolves to an array of annotations.
+     * @param {String} id URI for the targeted entity
+     * @param [String] targetStyle other formats of resource targeting.  May be null
+     */
+async function findByTargetId(id, targetStyle = [],queryUrl="http://tinydev.rerum.io/app/query") {
```
Comment (cubap) [inline js/entities.js:100]: straight copy

File: js/entities.js [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+
+/**
+ * Regularizes assertions on expanded objects to enforce the existence of a `source` key.
+ * The return is only the value of the assertion, so the desired key must be applied upstream
+ * from the scope of this function.
+ * @param any val asserted value of the incoming annotation.
+ * @param Object fromAnno parent annotation of the asserted value, as a handy metadata container.
+ * @returns Object with `value` and `source` keys.
+ */
+function buildValueObject(val, fromAnno) {
```
Comment (cubap) [inline js/entities.js:136]: cp

File: js/entities.js [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+        citationSource: fromAnno["@id"] || fromAnno.id,
+        citationNote: fromAnno.label || fromAnno.name || "Composed object from DEER",
+        comment: "Learn about the assembler for this object at https://github.com/CenterForDigitalHumanities/deer"
+    }
+    valueObject.value = val.value || getValue(val)
+    valueObject.evidence = val.evidence || fromAnno.evidence || ""
+    return valueObject
+}
+
+function getValue(property, alsoPeek = [], asType) {
```
Comment (cubap) [inline js/entities.js:148]: cp

File: index.html (hunk 1 above)
Comment (thehabes) [reply index.html:417]: Yeah we did a fresh release and i was wondering if you wanted to up the version.  We should do that today.

File: js/deer-utils.js (hunk 6 above)
Comment (thehabes) [reply js/deer-utils.js:119]: At some point I was having trouble with the scope of `this`.  Setting it at the top of the function strong-arms the scope to `UTILS` so that the key word `this` isn't conflated in loops and what not.  It may not be necessary any more, but if you plan to undo it for `this` make sure to test it and put breakpoints around the uses of `this` so you can be sure it is what you expect.

File: js/worker.js [hunk 11]
```diff
@@ -0,0 +1,162 @@
+const filesToCache = [
```
Comment (thehabes) [inline js/worker.js:1]: This is super fun

File: js/deer-record.js [hunk 12]
```diff
@@ -84,141 +75,42 @@ export default class DeerReport {
...
+            UTILS.worker.postMessage({
+                action: "form",
+                id: this.id
+            })
+            UTILS.worker.addEventListener("message", event => {
+                if (event.data.action === "expanded") {
+                    let obj = event.data.item
                     try {
                         let inputElems = this.inputs
-                        let flatKeys = inputElems.map(input => input.getAttribute(DEER.KEY))
```
Comment (cubap) [inline js/deer-record.js:91]: moved into utils.js
